tools/haskell_module_worker/src/Compile.hs

```diff
@@ -0,0 +1,279 @@
+    let (hs_srcs, non_hs_srcs) = partition isHaskellishTarget srcs
+    if null (non_hs_srcs) then do
+      hsc_env <- GHC.getSession
+      collectStatus logsRef $
+        mapM_ (compileFile hsc_env StopLn) hs_srcs
+    else
+      return $ NonHaskellInputs $ map fst non_hs_srcs
+
+-- | Creates a status from the exceptions thrown by the given action
+-- and the logs in the given 'IORef'.
+collectStatus :: IORef [String] -> IO () -> Ghc Status
+collectStatus logsRef action = do
+    hsc_env <- GHC.getSession
+    liftIO $ do
+      ee <- try action
+      logs <- reverse <$> readIORef logsRef
+      case ee of
+        Left se -> case fromException se of
+          Just e -> do
+            return $
+              CompileErrors logs $
+              map (showSDoc $ hsc_dflags hsc_env) $
+              pprErrMsgBagWithLoc $
+              srcErrorMessages e
+          Nothing -> case fromException se of
+            Just e | isPlainPanic e ->
+              return $
+                CompileErrors logs [show (e :: PlainGhcException)]
+            _ -> case fromException se of
+              Just e | not (isPanic e) ->
+                return $
+                  CompileErrors logs [show (e :: GhcException)]
+              _ ->
+                throwIO se
+        Right _ ->
+          return (Succeeded logs)
+  where
+    isPanic (Panic{}) = True
+    isPanic (PprPanic{}) = True
+    isPanic _ = False
+
+    isPlainPanic (PlainPanic{}) = True
+    isPlainPanic _ = False
+
+-- | Copied from @ghc/Main.hs@
+-- https://gitlab.haskell.org/ghc/ghc/-/blob/ghc-8.10.7-release/ghc/Main.hs#L297
+partition_args :: [String] -> [(String, Maybe Phase)] -> [String]
+               -> ([(String, Maybe Phase)], [String])
+partition_args [] srcs objs = (reverse srcs, reverse objs)
+partition_args ("-x":suff:args) srcs objs
+  | "none" <- suff      = partition_args args srcs objs
+  | StopLn <- phase     = partition_args args srcs (slurp ++ objs)
+  | otherwise           = partition_args rest (these_srcs ++ srcs) objs
+        where phase = startPhase suff
+              (slurp,rest) = break (== "-x") args
+              these_srcs = zip slurp (repeat (Just phase))
+partition_args (arg:args) srcs objs
+  | looks_like_an_input arg = partition_args args ((arg,Nothing):srcs) objs
+  | otherwise               = partition_args args srcs (arg:objs)
+
+looks_like_an_input :: String -> Bool
+looks_like_an_input m =  isSourceFilename m
+                      || looksLikeModuleName m
+                      || "-" `isPrefixOf` m
+                      || not (hasExtension m)
+
+-- Some of this function is copied from @compiler/main/DynFlags.hs@.
+-- https://gitlab.haskell.org/ghc/ghc/-/blob/ghc-8.10.7-release/compiler/main/DynFlags.hs#L2216
+renderLog
+  :: (String -> IO a)
+  -> DynFlags
+  -> WarnReason
+  -> Severity
+  -> SrcSpan
+  -> PprStyle
+  -> MsgDoc
+  -> IO a
+renderLog k dflags reason severity srcSpan pprStyle msg
+    = (>>= k) $ case severity of
```

Comment:
```suggestion
    = k =<< case severity of
```
Or is this part copied and perhaps not worth changing?

Reply:
Done.